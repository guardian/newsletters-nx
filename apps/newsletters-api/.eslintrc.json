{
	"extends": ["../../.eslintrc.json"],
	"ignorePatterns": ["!**/*"],
	"overrides": [
		{
			"files": ["*.ts", "*.tsx"],
			"parserOptions": {
				"project": ["apps/newsletters-api/tsconfig.*?.json"]
			},
			"rules": {}
		},
		{
			"files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
			"rules": {}
		},
		{
			"files": ["*.js", "*.jsx"],
			"rules": {}
		}
	]
}
