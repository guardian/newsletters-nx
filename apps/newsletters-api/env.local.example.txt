# Example values for .env.local file

# To use s3 locally (overriding the s3 values below, set to true)
USE_IN_MEMORY_STORAGE=false

# the name of the s3 bucket holding the data. (avoid committing actual bucket names in the repo)
NEWSLETTER_BUCKET_NAME=your-bucket-name

# aws region to use, probably 'eu-west-1' (default for Guardian accounts)
S3_REGIONS=eu-west-1

# the janus profile to use for credentials, probably 'developerPlayground' for local testing
S3_PROFILE=developerPlayground

# whether to serve the UI on the index route of the API
NEWSLETTERS_UI_SERVE=true

# required to determine the environment the app in running in
STAGE=DEV
STACK=newsletters
APP=newsletters-api

# toggle for whether to use dynamic image signing. This allow us to optionally sign uploaded image assets for email rendering
ENABLE_DYNAMIC_IMAGE_SIGNING=false

# toggle for whether to use the developer profile for local development
USE_DEVELOPER_PROFILE=true

# the name of the local user for development
LOCAL_USER_PROFILE_NAME="Software Developer"
# the Email Address of the local user for development
LOCAL_USER_PROFILE_EMAIL=software.developer@guardian.co.uk

# allows local running with developer permissions without needing to login
# Update the access level as required:
# 0 - Developer, 1 - Editor, 2 - Drafter, 3 - Viewer
USER_PERMISSIONS={"software.developer@guardian.co.uk": 0}

# whether to fetch from a local instance of email rendering on port 3010
# instead of the PROD service.
# USE_LOCAL_EMAIL_RENDERING=true

# The email addresses to which all emails are sent to when not on PROD.
# use semi-colons to separate the values.
# TEST_EMAIL_RECIPIENTS=firstname.lastname@guardian.co.uk;other.person@guardian.co.uk;
