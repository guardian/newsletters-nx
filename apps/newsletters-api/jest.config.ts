/* eslint-disable -- We want default export for config files */
export default {
	displayName: 'newsletters-api',
	preset: '../../jest.preset.js',
	globals: {},
	testEnvironment: 'node',
	transform: {
		'^.+\\.[tj]s$': [
			'ts-jest',
			{
				tsconfig: '<rootDir>/tsconfig.spec.json',
			},
		],
	},
	moduleFileExtensions: ['ts', 'js', 'html'],
	coverageDirectory: '../../coverage/apps/newsletters-api',
};
