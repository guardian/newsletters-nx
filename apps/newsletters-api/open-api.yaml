openapi: 3.0.3
info:
  title: Newsletters API
  description: The source of truth for newsletters at The Guardian.
  version: '0.1'
  contact:
    email: newsletters-dev@guardian.co.uk

paths:
  /legacy/newsletters:
    get:
      summary: Returns a list of newsletters.
      description: Provides the legacy API format, consisting only of the array of newsletters. Includes current and previous (cancelled or paused) editorial newsletters. Does not include marketing or product newsletters.
      responses:
        '200':
          description: A JSON array of newsletter entities
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: '#/components/schemas/LegacyNewsletter'

  /newsletters:
    get:
      summary: Returns a response object with a list of newsletters as the result field.
      description: Includes current and previous (cancelled or paused) editorial newsletters. Does not include marketing or product newsletters.
      responses:
        '200':
          description: A JSON object containing array of newsletter entities
          content:
            application/json:
              schema:
                type: object
                required:
                  - ok
                  - total
                  - results
                properties:
                  ok:
                    type: boolean
                    example: true
                  total:
                    type: number
                    example: 1
                    description: the number of newsletters in the results
                  results:
                    type: array
                    items:
                      # This will change to the "Newsletter" schema when the new data model is implemented
                      $ref: '#/components/schemas/LegacyNewsletter'

  /newsletters/{newsletterId}:
    get:
      summary: Get a newsletter by newsletterId
      parameters:
        - in: path
          name: newsletterId
          schema:
            type: string
          required: true
          description: The unique newsletterId of the newsletter to get
      responses:
        '404':
          description: A JSON object describing a 'not found' error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorResponse'
        '200':
          description: Returns a response object with the matching newsletters as the result field.
          content:
            application/json:
              schema:
                type: object
                required:
                  - results
                  - ok
                properties:
                  ok:
                    type: boolean
                    example: true
                  results:
                    # This will change to the "Newsletter" schema when the new data model is implemented
                    $ref: '#/components/schemas/LegacyNewsletter'

  /drafts:
    get:
      summary: Returns a response object with a list of drafts as the result field.
      description: Includes the objects representing draft newsletters yet to be launched.
      responses:
        '200':
          description: A JSON object containing array of draft entities
          content:
            application/json:
              schema:
                type: object
                required:
                  - ok
                  - total
                  - results
                properties:
                  ok:
                    type: boolean
                    example: true
                  total:
                    type: number
                    example: 1
                    description: the number of drafts in the results
                  results:
                    type: array
                    items:
                      $ref: '#/components/schemas/Draft'

  /drafts/{draftId}:
    get:
      summary: Get a draft by draftId
      parameters:
        - in: path
          name: draftId
          schema:
            type: number
          required: true
          description: The unique draftId number of the newsletter to get
      responses:
        '404':
          description: A JSON object describing a 'not found' error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorResponse'
        '200':
          description: Returns a response object with the matching draft as the result field.
          content:
            application/json:
              schema:
                type: object
                required:
                  - results
                  - ok
                properties:
                  ok:
                    type: boolean
                    example: true
                  results:
                    $ref: '#/components/schemas/Draft'

components:
  schemas:
    # TO DO - define schema for Newsletter when implemented in code.
    LegacyNewsletter:
      type: object
      required:
        - name
        - identityName
        - cancelled
        - theme
        - group
        - listId
        - listIdV1
        - emailEmbed
        - restricted
        - paused
        - emailConfirmation
      properties:
        identityName:
          type: string
          example: morning-briefing
        name:
          type: string
          example: First Edition
        cancelled:
          type: boolean
          example: false
        restricted:
          type: boolean
          example: false
        paused:
          type: boolean
          example: false
        emailConfirmation:
          type: boolean
          example: false
        brazeNewsletterName:
          type: string
          example: Editorial_MorningBriefingUK
        brazeSubscribeAttributeName:
          type: string
          example: MorningBriefingUk_Subscribe_Email
        brazeSubscribeEventNamePrefix:
          type: string
          example: morning_briefing_uk
        theme:
          type: string
          example: News
        group:
          type: string
          example: News in brief
        description:
          type: string
          example: Archie Bland and Nimo Omer take you through the top stories and what they mean, free every weekday morning
        regionFocus:
          type: string
          example: UK
        frequency:
          type: string
          example: Every weekday
        listIdV1:
          type: integer
          example: 3640
        listId:
          type: integer
          example: 4156
        exampleUrl:
          type: string
          example: https://www.theguardian.com/world/series/guardian-morning-briefing/latest/email
        signupPage:
          type: string
          example: https://www.theguardian.com/global/2022/sep/20/sign-up-for-the-first-edition-newsletter-our-free-news-email
        illustration:
          $ref: '#/components/schemas/NewsletterIllustration'
        emailEmbed:
          $ref: '#/components/schemas/EmailEmbed'
        campaignName:
          type: string
          example: MorningBriefingUK
        campaignCode:
          type: string
          example: morningbriefinguk_email
        brazeSubscribeAttributeNameAlternative:
          type: array
          items:
            type: string
            example: email_subscribe_morning_briefing_uk
    NewsletterIllustration:
      type: object
      required:
        - circle
      properties:
        circle:
          type: string
          example:
    EmailEmbed:
      type: object
      required:
        - name
        - title
        - successHeadline
        - successDescription
        - hexCode
      properties:
        name:
          type: string
          example: First Edition
        title:
          type: string
          example: Sign up for First Edition
        description:
          type: string
          example: Our morning email breaks down the key stories of the day, telling you whatâ€™s happening and why it matters
        successHeadline:
          type: string
          example: Subscription confirmed
        successDescription:
          type: string
          example: We'll send you First Edition every weekday
        hexCode:
          type: string
          example: '#DCDCDC'
        imageUrl:
          type: string
    ErrorResponse:
      type: object
      required:
        - ok
      properties:
        ok:
          type: boolean
          example: false
        message:
          type: string
          example: No newsletter with id fake-newsletter-id
    Draft:
      type: object
      required:
        - listId
      properties:
        listId:
          type: integer
          example: 7000
        name:
          type: string
          example: 'First edition'
