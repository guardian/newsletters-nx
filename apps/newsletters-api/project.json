{
	"name": "newsletters-api",
	"$schema": "../../node_modules/nx/schemas/project-schema.json",
	"sourceRoot": "apps/newsletters-api/src",
	"projectType": "application",
	"targets": {
		"build": {
			"executor": "@nrwl/esbuild:esbuild",
			"outputs": ["{options.outputPath}"],
			"options": {
				"outputPath": "dist/apps/newsletters-api",
				"format": ["cjs"],
				"main": "apps/newsletters-api/src/main.ts",
				"tsConfig": "apps/newsletters-api/tsconfig.app.json",
				"assets": ["apps/newsletters-api/src/assets"],
				"generatePackageJson": true
			}
		},
		"serve": {
			"executor": "@nrwl/js:node",
			"options": {
				"buildTarget": "newsletters-api:build"
			},
			"configurations": {
				"production": {
					"buildTarget": "newsletters-api:build:production"
				}
			}
		},
		"lint": {
			"executor": "@nrwl/linter:eslint",
			"outputs": ["{options.outputFile}"],
			"options": {
				"lintFilePatterns": ["apps/newsletters-api/**/*.ts"]
			}
		},
		"test": {
			"executor": "@nrwl/jest:jest",
			"outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
			"options": {
				"jestConfig": "apps/newsletters-api/jest.config.ts",
				"passWithNoTests": true
			}
		}
	},
	"tags": []
}
