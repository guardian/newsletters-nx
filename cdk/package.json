{
  "name": "cdk",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "test-update": "jest -u",
    "format": "prettier --write \"{lib,bin}/**/*.ts\"",
    "lint": "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
    "synth": "cdk synth --path-metadata false --version-reporting false",
    "diff": "cdk diff --path-metadata false --version-reporting false"
  },
  "devDependencies": {
    "@guardian/cdk": "49.0.3",
    "@guardian/eslint-config-typescript": "3.0.0",
    "@guardian/prettier": "1.0.0",
    "@types/jest": "^29.2.5",
    "@types/node": "18.11.18",
    "aws-cdk": "2.59.0",
    "aws-cdk-lib": "2.59.0",
    "constructs": "10.1.215",
    "eslint": "^8.32.0",
    "jest": "^29.3.1",
    "prettier": "^2.8.3",
    "source-map-support": "^0.5.20",
    "ts-jest": "^29.0.5",
    "ts-node": "^10.9.1",
    "typescript": "~4.9.4"
  },
  "prettier": "@guardian/prettier",
  "jest": {
    "testMatch": [
      "<rootDir>/lib/**/*.test.ts"
    ],
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "setupFilesAfterEnv": [
      "./jest.setup.js"
    ]
  },
  "eslintConfig": {
    "root": true,
    "env": {
      "node": true,
      "jest": true
    },
    "extends": [
      "@guardian/eslint-config-typescript"
    ],
    "parserOptions": {
      "ecmaVersion": 2020,
      "sourceType": "module"
    },
    "plugins": [
      "@typescript-eslint"
    ],
    "rules": {
      "@typescript-eslint/no-inferrable-types": 0,
      "import/no-namespace": 2
    },
    "ignorePatterns": [
      "**/*.js",
      "node_modules",
      "cdk.out",
      ".eslintrc.js",
      "jest.config.js"
    ]
  }
}