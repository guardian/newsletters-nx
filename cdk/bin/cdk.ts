import 'source-map-support/register';
import { App } from 'aws-cdk-lib';
import { Newsletters } from '../lib/newsletters';
import { NewslettersApi } from '../lib/newsletters-api';

const app = new App();

// Shared props for all CDK apps
export const sharedProps = {
	stack: 'newsletters',
	env: {
		region: 'eu-west-1',
	},
};

/** The internal tool for newsletter management */
const newslettersAppName = 'newsletters';

new Newsletters(app, 'Newsletters-CODE', {
	...sharedProps,
	stage: 'CODE',
	app: newslettersAppName,
	domainName: `${newslettersAppName}.code.dev-gutools.co.uk`,
});

// new NewslettersApi(app, 'NewslettersUi-PROD', {
// 	...sharedProps,
// 	stage: 'PROD',
// 	app: newslettersUiAppName,
// 	domainName: `${newslettersUiAppName}.gutools.co.uk`,
// });

/** The read-only newsletters API */
const newslettersApiAppName = 'newsletters-api';

new NewslettersApi(app, 'NewslettersApi-CODE', {
	...sharedProps,
	stage: 'CODE',
	app: newslettersApiAppName,
	/** @todo - this ought to be a .guardianapis.co.uk domain */
	domainName: `${newslettersApiAppName}.code.dev-gutools.co.uk`,
});

// new NewslettersApi(app, 'NewslettersApi-PROD', {
// 	...sharedProps,
// 	stage: 'PROD',
// 	app: newslettersApiAppName,
// /** @todo - this ought to be a .guardianapis.co.uk domain */
// 	domainName: `${newslettersApiAppName}.gutools.co.uk`,
// });
