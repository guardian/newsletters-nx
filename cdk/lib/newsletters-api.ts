import { GuNodeApp } from "@guardian/cdk";
import { AccessScope } from "@guardian/cdk/lib/constants";
import type { GuStackProps }from '@guardian/cdk/lib/constructs/core';
import { GuDistributionBucketParameter } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuS3Bucket } from "@guardian/cdk/lib/constructs/s3";
import type { App } from 'aws-cdk-lib';
import { InstanceClass, InstanceSize, InstanceType } from "aws-cdk-lib/aws-ec2";
import { StringParameter } from "aws-cdk-lib/aws-ssm"

export class NewslettersApi extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'newsletters-api'
		const bucketParameterName = `/${this.stage}/${this.stack}/${app}/s3BucketName`
		const bucketName = StringParameter.valueForStringParameter(this, bucketParameterName)

		new GuS3Bucket(this, "DataBucket", {
      bucketName,
      app,
      versioned: true,
    });

		const distributionBucketParameter = GuDistributionBucketParameter.getInstance(this)

		new GuNodeApp(this, {
			access: { scope: AccessScope.PUBLIC },
			certificateProps: {
				domainName: `newsletters-api.gutools.co.uk`
			},
			roleConfiguration: {
				// additionalPolicies: [
				// 	new GuAllowPolicy(this, "GetDistBucket", {
				// 		actions: ["s3:GetObject"],
				// 		resources: [`arn:aws:s3:::frontend-dist/*`],
				// 	}),
				// ],
			},
			monitoringConfiguration: {
				noMonitoring: true,
			},
			instanceType: InstanceType.of(
				InstanceClass.T4G,
				InstanceSize.MICRO
			),
			scaling: {
					minimumInstances: 1,
					maximumInstances: 2,
			},
			userData: [
				"#!/bin/bash",
				"set -e",
				"set +x",
				`aws s3 cp s3://${distributionBucketParameter.valueAsString}/${this.stack}/${this.stage}/${app}/index.js /tmp`,
				"chown ubuntu /tmp/index.js", // change ownership of the file
				"su ubuntu -c 'node /tmp/index.js'", // run the file as ubuntu user
			].join("\n"),
			app,
		})
	}
}
