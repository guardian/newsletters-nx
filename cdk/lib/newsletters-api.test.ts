import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { newslettersProps } from '../bin/cdk';
import { NewslettersApi } from './newsletters-api';

describe('The NewslettersApi stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new NewslettersApi(app, 'NewslettersApi-TEST', {
			...newslettersProps,
			stage: 'TEST',
			domainName: 'newsletters-api.test.dev-gutools.co.uk',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
