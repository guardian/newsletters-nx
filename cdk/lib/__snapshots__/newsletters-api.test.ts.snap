// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The NewslettersApi stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuS3Bucket",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "SsmParameterValueTESTnewslettersnewslettersapis3BucketNameC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/TEST/newsletters/newsletters-api/s3BucketName",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DataBucketNewslettersapi96EED09B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": {
          "Ref": "SsmParameterValueTESTnewslettersnewslettersapis3BucketNameC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters-api",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
